front/src/modules/simulationResult/components/SpaceTimeChart/useWaypointMenu.tsx

```diff
@@ -0,0 +1,71 @@
+import React, { useRef, useState } from 'react';
+
+import { EyeClosed } from '@osrd-project/ui-icons';
+import { useTranslation } from 'react-i18next';
+
+import type { OperationalPoint } from 'applications/operationalStudies/types';
+import type { OSRDMenuItem } from 'common/OSRDMenu';
+import useModalFocusTrap from 'utils/hooks/useModalFocusTrap';
+import useOutsideClick from 'utils/hooks/useOutsideClick';
+
+const SPACETIME_CHART_HEADER_HEIGHT = 40;
+const WAYPOINT_MENU_OFFSET = 2;
+
+const useWaypointMenu = (
+  manchetteWrapperRef: React.RefObject<HTMLDivElement>,
+  filteredWaypoints?: OperationalPoint[],
+  setFilteredWaypoints?: (waypoints: OperationalPoint[]) => void
+) => {
+  const { t } = useTranslation('simulation');
+
+  const [menuPosition, setMenuPosition] = useState<{ top: number; left: number }>();
+  const [activeOperationalPointId, setActiveOperationalPointId] = useState<string>();
+
+  const menuRef = useRef<HTMLDivElement>(null);
+
+  const closeMenu = () => {
+    setMenuPosition(undefined);
+    setActiveOperationalPointId(undefined);
+  };
+
+  useOutsideClick(menuRef, closeMenu);
+  useModalFocusTrap(menuRef, closeMenu);
+
+  const menuItems: OSRDMenuItem[] = [
+    {
+      title: t('waypointMenu.hide'),
+      icon: <EyeClosed />,
+      onClick: () => {
+        setMenuPosition(undefined);
+        setActiveOperationalPointId(undefined);
+        if (filteredWaypoints && setFilteredWaypoints) {
+          setFilteredWaypoints(
```

Comment:
This is racy, because another update to `filteredWaypoints` could be pending. Instead, `setFilteredWaypoints` can take a callback which mutates the current value:

```ts
setFilteredWaypoints((filteredWaypoints) => filteredWaypoints.filter(…));
```

As a bonus, this removes the need to pass `filteredWaypoints` as a prop.

Reply:
Yep I know but as it requires to pass the `setFilteredWaypoints` not as a standard function but as a `SetStateAction`, it needed to be changed everywhere we pass this props (but only two spots actually) and I was lazy :upside_down_face: But you're right I should do that, done :)